README.md
## About
This repository contains ROS packages for the WRL63-B robot arm

Operating System: Ubuntu 18.04, 20.04</br>
ROS Version: Melodic, Noetic

## Packages
rml_63_description
* Robot models and configuration files
* rm_63_bottom.urdf.xarco: RML63-B model without end tool

rml_63_moveit_config
* launch and configuration files to start MoveIt motion planning

rm_gazebo
* parameters and configuration files for Gazebo robot model

rml_63_demo
* MoveIt examples (obstacle avoidance, pick & play)

rm_msgs
* control and status messages

rm_control
* robot controller
* send cubic spline interpolated robot arm trajectories planned by MoveIt to rm_driver node every 20ms. Time period is adjustable but should be >10ms

rm_driver
* Establish socket connection with the robot arm through the ethernet port. Robot arm default IP address is 192.168.1.18
* subscribe to topics and update the joint angles of the robotic arm in RVIZ

rm_bringup
* automatically run rm_driver, rm_control, and RVIZ (MoveIt). Control real robot by dragging the simulated robot model

## Setup Workspace
1. Create workspace and change directory to the src folder
   ```
   $ mkdir -p <your_workspace>/src
   $ cd <your_workspace>/src
   ```
2. Git clone into the src folder
   ```
   git clone https://github.com/westonrobot/wrl63b_ros.git
   ```
3. Install system dependencies
   ```
   rosdep install -y --from-paths . --ignore-src --rosdistro <distro> -r
   ```
4. Build packages
   ```
   $ cd ..
   $ catkin init
   $ catkin build rm_msgs
   $ catkin build
   ```

## Visualize in RVIZ
Source robot workspace and launch rml_63_description package
```
roslaunch rml_63_description display_rml63_bottom.launch
```

If model is not shown, manually change the **Fixed Frame** to **base_link**, and add the robot model.
<img src="./docs/base_link.png"/>
<img src="./docs/robot_model.png"/>

## Motion Planning using MoveIt
MoveIt is a motion planning library that integrates component packages related to mobile operations in the ROS system. Through the MoveIt plug-in, you can control the robot arm to complete 
functions such as drag planning, random target planning, initial position and pose update, and collision detection.

Run MoveIt demo of the RML63B robotic arm in simulation
```
roslaunch rml_63_moveit_config demo.launch
```

1. Drag Planning
   * Drag the front end of the robot arm to change its position and pose. Then click the "Plan & Execute" button on the Planning tab
2. Target Planning
   * Click the Goal State drop-down list in the Planning tab to select the target position and pose of the robot arm, and then click the "Plan & Execute" button
   <img src="./docs/target_planning.png" />

## MoveIt with Gazebo Simulation
1. Modify the **rml_63_moveit_controller_manager.launch.xml** in the **rml_63_moveit_config** package to load the **controllers_gazebo.yaml** file 
<img src="./docs/moveit_controller_manager_sim.png"/>

2. Start RVIZ (MoveIt) and Gazebo
   ```
   roslaunch rm_gazebo rml_63_bringup_moveit.launch
   ```
3. Set **Fixed Frame** to **base_link**. Then add **MotionPlanning** into the panel. In RVIZ, drag the arm to a new position, then click **Plan & Execute**. The arm in Gazebo will then move to the position set in RVIZ. 

## Examples (Simulation)
1. Obstacles Avoidance
2. Pick & Place

### Obstacles Avoidance
1. Launch robotic arm in RVIZ with MoveIt
   ```
   roslaunch rml_63_moveit_config demo.launch
   ```
2. Launch python file containing obstacles in another terminal
   ```
   rosrun rml_63_demo moveit_obstacles_demo.py
   ```

   **Note**: Run "sudo chmod +x <python_file>.py" if the python files are not executable

### Pick & Place
1. Launch robotic arm in RVIZ with MoveIt
   ```
   roslaunch rml_63_moveit_config demo.launch
   ```
2. Launch pick_place_demo cpp file in another terminal
   ```
   rosrun rml_63_demo pick_place_demo
   ```

## MoveIt with real robot
1. Before operating it, make sure the robot and PC are connected to the same WiFi. Ping the robot and make sure there is a connection.
   ```
   ping 192.168.1.18
   ```
2. Modify the **rml_63_moveit_controller_manager.launch.xml** in the **rml_63_moveit_config** package to load the **controllers.yaml** file
<img src="./docs/moveit_controller_manager_real.png"/>

3. Launch rm_control package
   ```
   roslaunch rm_control rm_control.launch
   ```
3. Launch rm_driver package and MoveIt
   ```
   roslaunch rm_bringup rm_robot.launch
   ```
4. In RVIZ, drag the arm to a new position or select **Goal State**, then click **Plan & Execute**. The robotic arm will then move to the target location.

rml_63_robot/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package rm_robot
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.0 (2022-05-31)

------------------


rml_63_demo/src/test_api_movel.cpp
//
// Created by ubuntu on 21-9-14.
//
#include <ros/ros.h>
#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>
#include <moveit_msgs/AttachedCollisionObject.h>
#include <moveit_msgs/CollisionObject.h>

#include <tf/transform_listener.h>
#include <tf/transform_datatypes.h>
#include <tf/tf.h>
#include <rm_msgs/MoveL.h>
#include <rm_msgs/MoveC.h>
#include <rm_msgs/MoveJ.h>
#include <rm_msgs/Plan_State.h>
#include <math.h>
//#include <geometry_msgs/PointStamped.h>


// 接收到订阅的消息后，会进入消息回调函数
void planStateCallback(const rm_msgs::Plan_State::ConstPtr& msg)
{
    // 将接收到的消息打印出来
//    ROS_INFO("Marekr pose->position[%f, %f, %f]", msg->pose.position.x, msg->pose.position.y, msg->pose.position.z);
    if(msg->state)
    {
        ROS_INFO("*******Plan State OK");
    } else {
        ROS_INFO("*******Plan State Fail");
    }

}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "test_api_movel");
    ros::NodeHandle nh;

    ros::AsyncSpinner spin(3);
    spin.start();

    // 初始化需要使用move group控制的机械臂中的arm group
    moveit::planning_interface::MoveGroupInterface group("arm");
    // 设置位置(单位：米)和姿态（单位：弧度）的允许误差
    group.setGoalPositionTolerance(0.01);
    group.setGoalOrientationTolerance(0.02);


    /*
     * 1.相关初始化
     */
    //空间规划指令Publisher
//    ros::Publisher moveJ_pub = nh.advertise<rm_msgs::MoveJ>("/rm_driver/MoveJ_Cmd", 10);
    //直线规划指令Publisher
    ros::Publisher moveL_pub = nh.advertise<rm_msgs::MoveL>("/rm_driver/MoveL_Cmd", 10);


    ros::Subscriber planState_sub = nh.subscribe("/rm_driver/Plan_State", 10, planStateCallback);

    ros::Duration(2.0).sleep();

    /**
     * 2.控制机械臂运动到识别抓取的初始位姿
     */
    moveit::planning_interface::MoveGroupInterface::Plan mPlan;
    moveit::planning_interface::MoveItErrorCode planResult;


    //获取末端当前位姿
    geometry_msgs::PoseStamped currPose_ToolLink = group.getCurrentPose("link_flange");

    ROS_INFO("currPose_ToolLink.pose.position(%f, %f, %f)", currPose_ToolLink.pose.position.x, currPose_ToolLink.pose.position.y, currPose_ToolLink.pose.position.z);
    ROS_INFO("currPose_ToolLink.pose.orientation(%f, %f, %f, %f)", currPose_ToolLink.pose.orientation.x, currPose_ToolLink.pose.orientation.y, currPose_ToolLink.pose.orientation.z, currPose_ToolLink.pose.orientation.w);

    //tool_link目标位姿
    geometry_msgs::Pose targetPose_ToolLink;
    targetPose_ToolLink.position.x = currPose_ToolLink.pose.position.x;
    targetPose_ToolLink.position.y = currPose_ToolLink.pose.position.y + 0.04;
    targetPose_ToolLink.position.z = currPose_ToolLink.pose.position.z;
    targetPose_ToolLink.orientation = currPose_ToolLink.pose.orientation;

    ROS_INFO("targetPose_ToolLink.position(%f, %f, %f)", targetPose_ToolLink.position.x, targetPose_ToolLink.position.y, targetPose_ToolLink.position.z);
    ROS_INFO("targetPose_ToolLink.orientation(%f, %f, %f, %f)", targetPose_ToolLink.orientation.x, targetPose_ToolLink.orientation.y, targetPose_ToolLink.orientation.z, targetPose_ToolLink.orientation.w);


    rm_msgs::MoveL moveL_TargetPose;
    moveL_TargetPose.Pose = targetPose_ToolLink;
    moveL_TargetPose.speed = 0.2;
    moveL_pub.publish(moveL_TargetPose);


    ros::waitForShutdown();

    return 0;
}


/*
//定义一个MoveJ指令的起始位姿
rm_msgs::MoveJ moveJ_BeginPose;
moveJ_BeginPose.joint[0] = -0.360829;
moveJ_BeginPose.joint[1] = 0.528468;
moveJ_BeginPose.joint[2] = 1.326293;
moveJ_BeginPose.joint[3] = -0.000454;
moveJ_BeginPose.joint[4] = 1.221748;
moveJ_BeginPose.joint[5] = 0.000052;
moveJ_BeginPose.speed = 0.3;

//空间规划指令Publisher
moveJ_pub = nh.advertise<rm_msgs::MoveJ>("/rm_driver/MoveJ_Cmd", 10);

//发布空间规划指令运动到起始位姿
moveJ_pub.publish(moveJ_BeginPose);
*/


rml_63_demo/src/pick_place_demo.cpp
//
// Created by ubuntu on 21-8-11.
//
#include <ros/ros.h>
#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>
#include <moveit_msgs/AttachedCollisionObject.h>
#include <moveit_msgs/CollisionObject.h>


void addCollisionObjects(moveit::planning_interface::PlanningSceneInterface& planning_scene_interface)
{
    // 创建环境中的物体(两个桌子和一个抓取的目标物体)
    std::vector<moveit_msgs::CollisionObject> collision_objects;
    collision_objects.resize(3);

    // Add the first table where the cube will originally be kept.
    collision_objects[0].id = "table1";
    collision_objects[0].header.frame_id = "base_link";

    // 设置物体table1的形状及相关尺寸
    collision_objects[0].primitives.resize(1);
    collision_objects[0].primitives[0].type = collision_objects[0].primitives[0].BOX;
    collision_objects[0].primitives[0].dimensions.resize(3);
    collision_objects[0].primitives[0].dimensions[0] = 0.2;
    collision_objects[0].primitives[0].dimensions[1] = 0.4;
    collision_objects[0].primitives[0].dimensions[2] = 0.4;

    // 设置物体table1的位置
    collision_objects[0].primitive_poses.resize(1);
    collision_objects[0].primitive_poses[0].position.x = 0.45;
    collision_objects[0].primitive_poses[0].position.y = 0;
    collision_objects[0].primitive_poses[0].position.z = 0.2;

    collision_objects[0].operation = collision_objects[0].ADD;

    // Add the second table where we will be placing the cube.
    collision_objects[1].id = "table2";
    collision_objects[1].header.frame_id = "base_link";

    // 设置物体table2的形状及相关尺寸
    collision_objects[1].primitives.resize(1);
    collision_objects[1].primitives[0].type = collision_objects[1].primitives[0].BOX;
    collision_objects[1].primitives[0].dimensions.resize(3);
    collision_objects[1].primitives[0].dimensions[0] = 0.4;
    collision_objects[1].primitives[0].dimensions[1] = 0.2;
    collision_objects[1].primitives[0].dimensions[2] = 0.4;

    // 设置物体table2的位置
    collision_objects[1].primitive_poses.resize(1);
    collision_objects[1].primitive_poses[0].position.x = 0;
    collision_objects[1].primitive_poses[0].position.y = 0.45;
    collision_objects[1].primitive_poses[0].position.z = 0.2;

    collision_objects[1].operation = collision_objects[1].ADD;


    // 创建并配置一个抓取的目标物体
    collision_objects[2].header.frame_id = "base_link";
    collision_objects[2].id = "object";

    // 设置目标物体的形状和相关尺寸
    collision_objects[2].primitives.resize(1);
    collision_objects[2].primitives[0].type = collision_objects[1].primitives[0].BOX;
    collision_objects[2].primitives[0].dimensions.resize(3);
    collision_objects[2].primitives[0].dimensions[0] = 0.02;
    collision_objects[2].primitives[0].dimensions[1] = 0.02;
    collision_objects[2].primitives[0].dimensions[2] = 0.2;

    // 设置目标物体的位置
    collision_objects[2].primitive_poses.resize(1);
    collision_objects[2].primitive_poses[0].position.x = 0.44;
    collision_objects[2].primitive_poses[0].position.y = 0;
    collision_objects[2].primitive_poses[0].position.z = 0.5;

    collision_objects[2].operation = collision_objects[2].ADD;

    planning_scene_interface.applyCollisionObjects(collision_objects);
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "rm65_pick_place");
    ros::NodeHandle nh;
    ros::AsyncSpinner spin(1);
    spin.start();

//    ros::WallDuration(1.0).sleep();

    //创建运动规划的情景，等待创建完成
    moveit::planning_interface::PlanningSceneInterface planning_scene_interface;
    // 初始化需要使用move group控制的机械臂中的arm group
    moveit::planning_interface::MoveGroupInterface group("arm");
    // 设置位置(单位：米)和姿态（单位：弧度）的允许误差
    group.setGoalPositionTolerance(0.02);
    group.setGoalOrientationTolerance(0.05);
    sleep(5.0);

    // 在场景中创建相关物体
    addCollisionObjects(planning_scene_interface);

    // 等待一段时间来初始化ROS
    ros::WallDuration(1.0).sleep();

    // 设置机器人终端的目标位置
    geometry_msgs::Pose target_pose1;
//    target_pose1.orientation.w = 0.726282;
//    target_pose1.orientation.x= 4.04423e-07;
//    target_pose1.orientation.y = -0.687396;
//    target_pose1.orientation.z = 4.81813e-07;
//    target_pose1.orientation.w = 0.7;
//
//    target_pose1.position.x = 0.38;
//    target_pose1.position.y = 0;
//    target_pose1.position.z = 0.53;

    target_pose1.orientation.w = 0.703184;
    target_pose1.orientation.x= 1.42057e-05;
    target_pose1.orientation.y = 0.711007;
    target_pose1.orientation.z = -1.912e-05;

    target_pose1.position.x = 0.406138;
    target_pose1.position.y = 0.00100714;
    target_pose1.position.z = 0.478058;
    group.setPoseTarget(target_pose1, "link6");

    // 进行运动规划，计算机器人移动到目标的运动轨迹，此时只是计算出轨迹，并不会控制机械臂运动
    moveit::planning_interface::MoveGroupInterface::Plan my_plan;
    moveit::planning_interface::MoveItErrorCode success = group.plan(my_plan);

    ROS_INFO("Visualizing plan 1 (pose goal) %s",success?"target_pose1":"FAILED");

    //让机械臂按照规划的轨迹开始运动。
    if(success)
    {
        group.execute(my_plan);

        ros::WallDuration(1.0).sleep();

        //创建一个附加到Link6上的物体
        moveit_msgs::AttachedCollisionObject attached_object;
        attached_object.link_name = "link6";
        /* The header must contain a valid TF frame*/
        attached_object.object.header.frame_id = "base_link";
        /* 设置物体ID */
        attached_object.object.id = "object1";

        /* 设置目标物体的位置 */
        geometry_msgs::Pose pose;
//    pose.orientation.w = 1.0;
        pose.position.x = 0.44;
        pose.position.y = 0;
        pose.position.z = 0.5;

        /* 定义物体的形状和相关尺寸 */
        shape_msgs::SolidPrimitive primitive;
        primitive.type = primitive.BOX;
        primitive.dimensions.resize(3);
        primitive.dimensions[0] = 0.02;
        primitive.dimensions[1] = 0.02;
        primitive.dimensions[2] = 0.2;

        attached_object.object.primitives.push_back(primitive);
        attached_object.object.primitive_poses.push_back(pose);

        // 将一个对象附加到机器人要指定为ADD操作的对应操作
        attached_object.object.operation = attached_object.object.ADD;

        // 因为要将物体连接到机器人的手上以模拟拿起物体所以配置碰撞检查器忽略物体和机器人手之间的碰撞
        attached_object.touch_links = std::vector<std::string>{"link6"};

        // 将创建的附加物体应用到场景中
        planning_scene_interface.applyAttachedCollisionObject(attached_object);
        // 从场景中移除原来的目标物体
        planning_scene_interface.removeCollisionObjects(std::vector<std::string>{"object"});

//    ros::WallDuration(1.0).sleep();

        // 设置放置点的目标位置
        geometry_msgs::Pose target_pose2;
        target_pose2.orientation.w = 0.501046;
        target_pose2.orientation.x= -0.49893;
        target_pose2.orientation.y = 0.49938;
        target_pose2.orientation.z = 0.500641;

        target_pose2.position.x = -0.000392824;
        target_pose2.position.y = 0.394717;
        target_pose2.position.z = 0.504879;

//    target_pose2.orientation.w = 0.500196;
//    target_pose2.orientation.x= -0.49973;
//    target_pose2.orientation.y = 0.500227;
//    target_pose2.orientation.z = 0.499847;
//
//    target_pose2.position.x = -0.000394317;
//    target_pose2.position.y = 0.401346;
//    target_pose2.position.z = 0.467557;

//    target_pose2.orientation.w = 0.498372;
//    target_pose2.orientation.x= -0.501688;
//    target_pose2.orientation.y = 0.50205;
//    target_pose2.orientation.z = 0.497876;
//
//    target_pose2.position.x = -0.00039465;
//    target_pose2.position.y = 0.394717;
//    target_pose2.position.z = 0.504879;
        group.setPoseTarget(target_pose2, "link6");

        // 进行运动规划，计算机器人移动到目标的运动轨迹，此时只是计算出轨迹，并不会控制机械臂运动
        moveit::planning_interface::MoveGroupInterface::Plan my_plan2;
        moveit::planning_interface::MoveItErrorCode success2 = group.plan(my_plan2);

        ROS_INFO("Visualizing plan 2 (pose goal) %s",success2?"target_pose2":"FAILED");

        //让机械臂按照规划的轨迹开始运动。
        if(success2)
        {
            group.execute(my_plan2);

            ros::WallDuration(1.0).sleep();

            //创建一个在放置点处的物体
            std::vector<moveit_msgs::CollisionObject> collision_objects1;
            collision_objects1.resize(1);

            collision_objects1[0].id = "object2";
            collision_objects1[0].header.frame_id = "base_link";

            // 设置物体的形状及相关尺寸
            collision_objects1[0].primitives.resize(1);
            collision_objects1[0].primitives[0].type = collision_objects1[0].primitives[0].BOX;
            collision_objects1[0].primitives[0].dimensions.resize(3);
            collision_objects1[0].primitives[0].dimensions[0] = 0.02;
            collision_objects1[0].primitives[0].dimensions[1] = 0.02;
            collision_objects1[0].primitives[0].dimensions[2] = 0.2;

            // 设置物体位置
            collision_objects1[0].primitive_poses.resize(1);
            collision_objects1[0].primitive_poses[0].position.x = 0;
            collision_objects1[0].primitive_poses[0].position.y = 0.44;
            collision_objects1[0].primitive_poses[0].position.z = 0.5;

            collision_objects1[0].operation = collision_objects1[0].ADD;


            //创建一个用于解除附加到机械臂Link6的附加物体
            moveit_msgs::AttachedCollisionObject detach_object;
            detach_object.link_name = "link6";
            /* The header must contain a valid TF frame*/
            detach_object.object.header.frame_id = "base_link";
            /* The id of the object */
            detach_object.object.id = "object1";
            detach_object.object.operation = attached_object.object.REMOVE;

            //达到放置点后移除附加物体然后在放置点加入放置的物体
            planning_scene_interface.applyAttachedCollisionObject(detach_object);
            planning_scene_interface.removeCollisionObjects(std::vector<std::string>{"object1"});
            planning_scene_interface.addCollisionObjects(collision_objects1);

            //设置机械臂返回zero姿态
            group.setNamedTarget("zero");
            // 进行运动规划，计算机器人移动到目标的运动轨迹，此时只是计算出轨迹，并不会控制机械臂运动
            moveit::planning_interface::MoveGroupInterface::Plan my_plan3;
            moveit::planning_interface::MoveItErrorCode success3 = group.plan(my_plan3);

            ROS_INFO("Visualizing plan 3 (pose goal) %s",success3?"zero":"FAILED");

            //让机械臂按照规划的轨迹开始运动。
            if(success3)
                group.execute(my_plan3);
        }
    }

    ros::waitForShutdown();

    return 0;
}


rml_63_demo/scripts/moveit_obstacles_demo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import rospy, sys
import moveit_commander
from moveit_commander import MoveGroupCommander, PlanningSceneInterface
from moveit_msgs.msg import  PlanningScene, ObjectColor
from geometry_msgs.msg import PoseStamped, Pose

class MoveItObstaclesDemo:
    def __init__(self):
        # 初始化move_group的API
        moveit_commander.roscpp_initialize(sys.argv)
        
        # 初始化ROS节点
        rospy.init_node('moveit_obstacles_demo')
        
        # 初始化场景对象
        scene = PlanningSceneInterface()
        
        # 创建一个发布场景变化信息的发布者
        self.scene_pub = rospy.Publisher('planning_scene', PlanningScene, queue_size=5)
        
        # 创建一个存储物体颜色的字典对象
        self.colors = dict()
        
        # 等待场景准备就绪
        rospy.sleep(1)
                        
        # 初始化需要使用move group控制的机械臂中的arm group
        arm = MoveGroupCommander('arm')
        
        # 获取终端link的名称
        end_effector_link = arm.get_end_effector_link()
        
        # 设置位置(单位：米)和姿态（单位：弧度）的允许误差
        arm.set_goal_position_tolerance(0.01)
        arm.set_goal_orientation_tolerance(0.05)
       
        # 当运动规划失败后，允许重新规划
        arm.allow_replanning(True)
        
        # 设置目标位置所使用的参考坐标系
        reference_frame = 'base_link'
        arm.set_pose_reference_frame(reference_frame)
        
        # 设置每次运动规划的时间限制：5s
        arm.set_planning_time(5)
        
        # 设置场景物体的名称
        table_id = 'table'
        box1_id = 'box1'
        box2_id = 'box2'
        
        # 移除场景中之前运行残留的物体
        scene.remove_world_object(table_id)
        scene.remove_world_object(box1_id)
        scene.remove_world_object(box2_id)    
        rospy.sleep(1)
        
        # 控制机械臂先回到初始化位置
        arm.set_named_target('zero')
        arm.go()
        rospy.sleep(2)
        
        # 设置桌面的高度
        table_ground = 0.35

        # 设置table、box1和box2的三维尺寸
        # table_size = [0.2, 0.7, 0.01]
        table_size = [0.8, 0.2, 0.03]
        box1_size = [0.1, 0.05, 0.1]
        box2_size = [0.05, 0.05, 0.3]

        # 将三个物体加入场景当中
        table_pose = PoseStamped()
        table_pose.header.frame_id = reference_frame
        table_pose.pose.position.x = 0.0
        table_pose.pose.position.y = 0.33
        # table_pose.pose.position.z = table_ground + table_size[2] / 2.0
        table_pose.pose.position.z = 0.65
        table_pose.pose.orientation.w = 1.0
        scene.add_box(table_id, table_pose, table_size)
        # rospy.sleep(2)

        # box1_pose = PoseStamped()
        # box1_pose.header.frame_id = reference_frame
        # box1_pose.pose.position.x = 0.31
        # box1_pose.pose.position.y = -0.25
        # box1_pose.pose.position.z = table_ground + table_size[2] + box1_size[2] / 2.0
        # box1_pose.pose.orientation.w = 1.0
        # scene.add_box(box1_id, box1_pose, box1_size)
        # # rospy.sleep(2)
        #
        # box2_pose = PoseStamped()
        # box2_pose.header.frame_id = reference_frame
        # box2_pose.pose.position.x = 0.29
        # box2_pose.pose.position.y = 0.02
        # box2_pose.pose.position.z = table_ground + table_size[2] + box2_size[2] / 2.0
        # box2_pose.pose.orientation.w = 1.0
        # scene.add_box(box2_id, box2_pose, box2_size)
        # rospy.sleep(2)

        # 将桌子设置成红色，两个box设置成橙色
        # self.setColor(table_id, 0.8, 0, 0, 1.0)
        # self.setColor(box1_id, 0.8, 0.4, 0, 1.0)
        # self.setColor(box2_id, 0.8, 0.4, 0, 1.0)

        # 将场景中的颜色设置发布
        # self.sendColors()
        # rospy.sleep(2)

        # 控制机械臂回到初始化位置
        arm.set_named_target('forward')
        arm.go()
        rospy.sleep(2)

        # 设置机械臂的运动目标位置，位于桌面之上两个盒子之间
        # target_pose = PoseStamped()
        # target_pose.header.frame_id = reference_frame
        # target_pose.pose.position.x = 0.32
        # target_pose.pose.position.y = -0.22
        # # target_pose.pose.position.z = table_pose.pose.position.z + table_size[2] + 0.05
        # target_pose.pose.position.z = 0.42
        # target_pose.pose.orientation.w = 0.0
        #
        # # 控制机械臂运动到目标位置
        # arm.set_pose_target(target_pose, end_effector_link)
        # arm.go()
        # rospy.sleep(2)

        # # 设置机械臂的运动目标位置，进行避障规划
        # target_pose2 = PoseStamped()
        # target_pose2.header.frame_id = reference_frame
        # target_pose2.pose.position.x = 0.2
        # target_pose2.pose.position.y = -0.25
        # target_pose2.pose.position.z = table_pose.pose.position.z + table_size[2] + 0.05
        # target_pose2.pose.orientation.w = 1.0
        #
        # # 控制机械臂运动到目标位置
        # arm.set_pose_target(target_pose2, end_effector_link)
        # arm.go()
        # rospy.sleep(2)
        #
        # 控制机械臂回到初始化位置
        arm.set_named_target('zero')
        arm.go()
        
        # 关闭并退出moveit
        moveit_commander.roscpp_shutdown()
        moveit_commander.os._exit(0)
        
    # 设置场景物体的颜色
    def setColor(self, name, r, g, b, a = 0.9):
        # 初始化moveit颜色对象
        color = ObjectColor()
        
        # 设置颜色值
        color.id = name
        color.color.r = r
        color.color.g = g
        color.color.b = b
        color.color.a = a
        
        # 更新颜色字典
        self.colors[name] = color

    # 将颜色设置发送并应用到moveit场景当中
    def sendColors(self):
        # 初始化规划场景对象
        p = PlanningScene()

        # 需要设置规划场景是否有差异     
        p.is_diff = True
        
        # 从颜色字典中取出颜色设置
        for color in self.colors.values():
            p.object_colors.append(color)
        
        # 发布场景物体颜色设置
        self.scene_pub.publish(p)

if __name__ == "__main__":
    try:
        MoveItObstaclesDemo()
    except KeyboardInterrupt:
        raise
    

    
